CHAID/stats.py

```diff
@@ -73,40 +77,103 @@ def best_cat_split(self, ind, dep):
+            break_out = False
+            has_merged = 1
+            while has_merged == 1:
+                has_merged = 0
+                data = np.array([ [k, sum(v.values())] for k,v in freq.items() ])
+                data_sorted = data[data[:,1].argsort()]
+                # break out of they will never make enough sample, or there's two and any are too small
+                if sum(data_sorted[:, 1]) < self.min_child_node_size or ( len(data_sorted[:, 1]) == 2 and (data_sorted[0:2, 1] < self.min_child_node_size).any()):
+                    break_out = True
+                    break
+
+                if (data_sorted[0:2, 1] < self.min_child_node_size).all():
+                    # want [1, 2, 3] not [3, 1, 2]
+                    first = min(data_sorted[0, 0], data_sorted[1, 0])
+                    second = max(data_sorted[0, 0], data_sorted[1, 0])
+                    ind_var.group(first, second)
+                    for val, count in freq[second].items():
+                        freq[first][val] += count
+                    del freq[second]
+                    has_merged = 1
+
+            if break_out == True:
+                split.invalid_reason = InvalidSplitReason.MIN_CHILD_NODE_SIZE
+                continue
+
             if len(list(ind_var.possible_groupings())) == 0:
                 split.invalid_reason = InvalidSplitReason.PURE_NODE
 
-            choice, highest_p_join, split_chi, dof = None, None, None, None
-            for comb in ind_var.all_combinations():
-                freqs = [ sum( [ cl.Counter(freq[key]) for key in c ], cl.Counter()) for c in comb ]
-                keys = set(sum([ list(f.keys()) for f in freqs ], []))
+            while next(ind_var.possible_groupings(), None) is not None:
+                choice, highest_p_join, split_chi = None, None, None
+
+                for comb in ind_var.possible_groupings():
+                    col1_freq = freq[comb[0]]
+                    col2_freq = freq[comb[1]]
+
+                    keys = set(col1_freq.keys()).union(col2_freq.keys())
+
+                    n_ij = np.array([
+                        [col1_freq.get(k, 0) for k in keys],
+                        [col2_freq.get(k, 0) for k in keys]
+                    ])
+
+                    if n_ij.shape[1] == 1:
+                        p_split, dof, chi = 1, float('nan'), float('nan')
+                        break
+                    else:
+                        chi, p_split, dof = chisquare(n_ij, dep.weights is not None)
+
+                    if choice is None or p_split > highest_p_join or (p_split == highest_p_join and chi > split_chi):
+                        choice, highest_p_join, split_chi = comb, p_split, chi
+
+                invalid_reason = None
+                sufficient_split = highest_p_join < self.alpha_merge
+                if not sufficient_split: invalid_reason = InvalidSplitReason.ALPHA_MERGE
```

Comment:
I think dropping the `invalid_reason = None` and using `if ... elif .. else` is cleaner here

```python
if highest_p_join >= self.alpha_merge:
  split.invalid_reason = InvalidSplitReason.ALPHA_MERGE
elif any( sum(node_v.values()) < self.min_child_node_size for node_v in freq.values() ):
  split.invalid_reason = InvalidSplitReason.MIN_CHILD_NODE_SIZE
elif sufficient_split and len(freq.values()) > 1:
  # Merge loop here
```